src/libcore/cell.rs

```diff
@@ -1052,7 +1056,7 @@ impl<'b> Clone for BorrowRef<'b> {
         debug_assert!(borrow != UNUSED);
         // Prevent the borrow counter from overflowing into
         // a writing borrow.
-        assert!(borrow < MIN_WRITING - 1);
+        assert!(borrow != isize::max_value());
```

Comment:
Hm but what if `borrow` is `isize::min_value()` because it overflowed above? Then this will count upwards towards 0.

Seems like by first calling `borrow()` often enough to get the flag to overflow, and then calling `map_split` often enough to get the counter back to 0, I can break this?

Is that one check for making sure `BorrowRef::new` doesn't overflow really so expensive?

Reply:
At minimum the above comment ("Since this Ref exists, we know the borrow flag is not set to WRITING") needs to be updated.

I dont' see how map_split is relevant, do you mean calling `borrow()` often enough to overflow to a negative number and then calling `Clone` often enough to get the counter back to zero? That's a good point, but I think it can be fixed by doing a < check like in `borrow` (either directly here, at the loss of 1/2^31 possible RefMuts, or by incrementing first and then checking it's still positive). All other locations that increment the counter need similar checks.

Reply:
This is, I believe, a mistake. Now that we're comfortable with overflowing into the writing territory, we should just remove this assertion.

Reply:
Removing the check entirely would be wrong. Even if we're fine with overflowing past `isize::MAX`, we need to make sure we're not going much further, to 0, because that would allow having a `Ref` alongside a `RefMut` and thus be unsound.

Reply:
Ah, I see. That's a fair point. I was thinking that it wouldn't be possible to create a new read reference once the refcount was in writing territory, but of course that's only true if you're actually checking first, which happens in `BorrowRef::new`, but not in `clone`.

Reply:
Regardless, I recently saw the discussion about whether to allow overflow at all, and I saw that the conclusion was not to do it, so I'm going to go back to disallowing it.